File: packages/runar-go/sdk_woc_provider.go
package runar

import (
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"strings"

	"github.com/bsv-blockchain/go-sdk/transaction"
)

// ---------------------------------------------------------------------------
// WhatsOnChainProvider — HTTP-based BSV API provider
// ---------------------------------------------------------------------------

// WhatsOnChainProvider implements the Provider interface using the
// WhatsOnChain REST API (https://whatsonchain.com).
type WhatsOnChainProvider struct {
	Network string // "mainnet" or "testnet"
	baseURL string
	client  *http.Client
}

// NewWhatsOnChainProvider creates a new WhatsOnChainProvider for the given network.
// Network must be "mainnet" or "testnet" (defaults to "mainnet" if empty).
func NewWhatsOnChainProvider(network string) *WhatsOnChainProvider {
	if network == "" {
		network = "mainnet"
	}
	baseURL := "https://api.whatsonchain.com/v1/bsv/main"
	if network == "testnet" {
		baseURL = "https://api.whatsonchain.com/v1/bsv/test"
	}
	return &WhatsOnChainProvider{
		Network: network,
		baseURL: baseURL,
		client:  &http.Client{},
	}
}

// ---------------------------------------------------------------------------
// WoC API response shapes
// ---------------------------------------------------------------------------

type wocTxVin struct {
	Txid      string `json:"txid"`
	Vout      int    `json:"vout"`
	ScriptSig struct {
		Hex string `json:"hex"`
	} `json:"scriptSig"`
	Sequence uint32 `json:"sequence"`
}

type wocTxVout struct {
	Value        float64 `json:"value"`
	N            int     `json:"n"`
	ScriptPubKey struct {
		Hex string `json:"hex"`
	} `json:"scriptPubKey"`
}

type wocTxResponse struct {
	Txid     string      `json:"txid"`
	Version  int         `json:"version"`
	Vin      []wocTxVin  `json:"vin"`
	Vout     []wocTxVout `json:"vout"`
	Locktime int         `json:"locktime"`
	Hex      string      `json:"hex,omitempty"`
}

type wocUtxoEntry struct {
	TxHash string `json:"tx_hash"`
	TxPos  int    `json:"tx_pos"`
	Value  int64  `json:"value"`
	Height int    `json:"height"`
}

// ---------------------------------------------------------------------------
// Provider interface implementation
// ---------------------------------------------------------------------------

// GetTransaction fetches a transaction by its txid from WhatsOnChain.
func (p *WhatsOnChainProvider) GetTransaction(txid string) (*TransactionData, error) {
	url := fmt.Sprintf("%s/tx/hash/%s", p.baseURL, txid)
	resp, err := p.client.Get(url)
	if err != nil {
		return nil, fmt.Errorf("WoC getTransaction request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("WoC getTransaction failed (%d): %s", resp.StatusCode, string(body))
	}

	var data wocTxResponse
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, fmt.Errorf("WoC getTransaction JSON decode failed: %w", err)
	}

	inputs := make([]TxInput, len(data.Vin))
	for i, vin := range data.Vin {
		inputs[i] = TxInput{
			Txid:        vin.Txid,
			OutputIndex: vin.Vout,
			Script:      vin.ScriptSig.Hex,
			Sequence:    vin.Sequence,
		}
	}

	outputs := make([]TxOutput, len(data.Vout))
	for i, vout := range data.Vout {
		outputs[i] = TxOutput{
			Satoshis: int64(math.Round(vout.Value * 1e8)),
			Script:   vout.ScriptPubKey.Hex,
		}
	}

	return &TransactionData{
		Txid:     data.Txid,
		Version:  data.Version,
		Inputs:   inputs,
		Outputs:  outputs,
		Locktime: data.Locktime,
		Raw:      data.Hex,
	}, nil
}

// Broadcast sends a transaction to the network via WhatsOnChain.
// Returns the txid on success.
func (p *WhatsOnChainProvider) Broadcast(tx *transaction.Transaction) (string, error) {
	rawTx := tx.Hex()
	payload := fmt.Sprintf(`{"txhex":"%s"}`, rawTx)
	resp, err := p.client.Post(
		fmt.Sprintf("%s/tx/raw", p.baseURL),
		"application/json",
		strings.NewReader(payload),
	)
	if err != nil {
		return "", fmt.Errorf("WoC broadcast request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("WoC broadcast failed (%d): %s", resp.StatusCode, string(body))
	}

	// WoC returns the txid as a JSON-encoded string
	var txid string
	if err := json.NewDecoder(resp.Body).Decode(&txid); err != nil {
		return "", fmt.Errorf("WoC broadcast JSON decode failed: %w", err)
	}
	return txid, nil
}

// GetUtxos returns all UTXOs for a given address from WhatsOnChain.
func (p *WhatsOnChainProvider) GetUtxos(address string) ([]UTXO, error) {
	url := fmt.Sprintf("%s/address/%s/unspent", p.baseURL, address)
	resp, err := p.client.Get(url)
	if err != nil {
		return nil, fmt.Errorf("WoC getUtxos request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("WoC getUtxos failed (%d): %s", resp.StatusCode, string(body))
	}

	var entries []wocUtxoEntry
	if err := json.NewDecoder(resp.Body).Decode(&entries); err != nil {
		return nil, fmt.Errorf("WoC getUtxos JSON decode failed: %w", err)
	}

	utxos := make([]UTXO, len(entries))
	for i, e := range entries {
		utxos[i] = UTXO{
			Txid:        e.TxHash,
			OutputIndex: e.TxPos,
			Satoshis:    e.Value,
			Script:      "", // WoC doesn't return locking script in UTXO list
		}
	}
	return utxos, nil
}

// GetContractUtxo finds a UTXO by its script hash from WhatsOnChain.
// Returns nil if no UTXO is found with the given script hash.
func (p *WhatsOnChainProvider) GetContractUtxo(scriptHash string) (*UTXO, error) {
	url := fmt.Sprintf("%s/script/%s/unspent", p.baseURL, scriptHash)
	resp, err := p.client.Get(url)
	if err != nil {
		return nil, fmt.Errorf("WoC getContractUtxo request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return nil, nil
	}
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("WoC getContractUtxo failed (%d): %s", resp.StatusCode, string(body))
	}

	var entries []wocUtxoEntry
	if err := json.NewDecoder(resp.Body).Decode(&entries); err != nil {
		return nil, fmt.Errorf("WoC getContractUtxo JSON decode failed: %w", err)
	}

	if len(entries) == 0 {
		return nil, nil
	}

	first := entries[0]
	return &UTXO{
		Txid:        first.TxHash,
		OutputIndex: first.TxPos,
		Satoshis:    first.Value,
		Script:      "",
	}, nil
}

// GetNetwork returns the network this provider is connected to.
func (p *WhatsOnChainProvider) GetNetwork() string {
	return p.Network
}

// GetRawTransaction fetches the raw transaction hex by its txid.
func (p *WhatsOnChainProvider) GetRawTransaction(txid string) (string, error) {
	url := fmt.Sprintf("%s/tx/%s/hex", p.baseURL, txid)
	resp, err := p.client.Get(url)
	if err != nil {
		return "", fmt.Errorf("WoC getRawTransaction request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("WoC getRawTransaction failed (%d): %s", resp.StatusCode, string(body))
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("WoC getRawTransaction read failed: %w", err)
	}
	return strings.TrimSpace(string(body)), nil
}

// GetFeeRate returns the fee rate in satoshis per KB.
// BSV standard relay fee is 0.1 sat/byte (100 sat/KB).
func (p *WhatsOnChainProvider) GetFeeRate() (int64, error) {
	return 100, nil
}

File: packages/runar-zig/src/sdk_woc_provider.zig
const std = @import("std");
const types = @import("sdk_types.zig");
const provider_mod = @import("sdk_provider.zig");

// ---------------------------------------------------------------------------
// WhatsOnChainProvider — HTTP-based BSV blockchain provider via WoC API
// ---------------------------------------------------------------------------

/// WhatsOnChainProvider implements Provider by making HTTP requests to the
/// WhatsOnChain API (https://whatsonchain.com). Supports mainnet and testnet.
pub const WhatsOnChainProvider = struct {
    allocator: std.mem.Allocator,
    network: Network,
    base_url: []const u8,

    pub const Network = enum {
        mainnet,
        testnet,

        pub fn toString(self: Network) []const u8 {
            return switch (self) {
                .mainnet => "mainnet",
                .testnet => "testnet",
            };
        }
    };

    const mainnet_base = "https://api.whatsonchain.com/v1/bsv/main";
    const testnet_base = "https://api.whatsonchain.com/v1/bsv/test";

    pub fn init(allocator: std.mem.Allocator, network: Network) WhatsOnChainProvider {
        return .{
            .allocator = allocator,
            .network = network,
            .base_url = switch (network) {
                .mainnet => mainnet_base,
                .testnet => testnet_base,
            },
        };
    }

    pub fn deinit(self: *WhatsOnChainProvider) void {
        _ = self;
    }

    /// Return a Provider interface backed by this WhatsOnChainProvider.
    pub fn provider(self: *WhatsOnChainProvider) provider_mod.Provider {
        return .{
            .ptr = @ptrCast(self),
            .vtable = &vtable,
        };
    }

    const vtable = provider_mod.Provider.VTable{
        .getTransaction = getTransactionImpl,
        .broadcast = broadcastImpl,
        .getUtxos = getUtxosImpl,
        .getContractUtxo = getContractUtxoImpl,
        .getNetwork = getNetworkImpl,
        .getFeeRate = getFeeRateImpl,
        .getRawTransaction = getRawTransactionImpl,
    };

    // -----------------------------------------------------------------------
    // HTTP helper
    // -----------------------------------------------------------------------

    fn httpGet(self: *WhatsOnChainProvider, allocator: std.mem.Allocator, path: []const u8) provider_mod.ProviderError![]u8 {
        const url = std.fmt.allocPrint(self.allocator, "{s}{s}", .{ self.base_url, path }) catch return provider_mod.ProviderError.OutOfMemory;
        defer self.allocator.free(url);

        var client = std.http.Client{ .allocator = allocator };
        defer client.deinit();

        var buf: [8192]u8 = undefined;
        const uri = std.Uri.parse(url) catch return provider_mod.ProviderError.NetworkError;
        var req = client.open(.GET, uri, .{ .server_header_buffer = &buf }) catch return provider_mod.ProviderError.NetworkError;
        defer req.deinit();

        req.send() catch return provider_mod.ProviderError.NetworkError;
        req.wait() catch return provider_mod.ProviderError.NetworkError;

        if (req.status != .ok) {
            if (req.status == .not_found) return provider_mod.ProviderError.NotFound;
            return provider_mod.ProviderError.NetworkError;
        }

        const body = req.reader().readAllAlloc(allocator, 10 * 1024 * 1024) catch return provider_mod.ProviderError.OutOfMemory;
        return body;
    }

    fn httpPost(self: *WhatsOnChainProvider, allocator: std.mem.Allocator, path: []const u8, json_body: []const u8) provider_mod.ProviderError![]u8 {
        const url = std.fmt.allocPrint(self.allocator, "{s}{s}", .{ self.base_url, path }) catch return provider_mod.ProviderError.OutOfMemory;
        defer self.allocator.free(url);

        var client = std.http.Client{ .allocator = allocator };
        defer client.deinit();

        var buf: [8192]u8 = undefined;
        const uri = std.Uri.parse(url) catch return provider_mod.ProviderError.NetworkError;
        var req = client.open(.POST, uri, .{
            .server_header_buffer = &buf,
            .extra_headers = &.{
                .{ .name = "Content-Type", .value = "application/json" },
            },
        }) catch return provider_mod.ProviderError.NetworkError;
        defer req.deinit();

        req.transfer_encoding = .{ .content_length = json_body.len };
        req.send() catch return provider_mod.ProviderError.NetworkError;
        req.writeAll(json_body) catch return provider_mod.ProviderError.NetworkError;
        req.finish() catch return provider_mod.ProviderError.NetworkError;
        req.wait() catch return provider_mod.ProviderError.NetworkError;

        if (req.status != .ok) {
            return provider_mod.ProviderError.BroadcastFailed;
        }

        const body = req.reader().readAllAlloc(allocator, 10 * 1024 * 1024) catch return provider_mod.ProviderError.OutOfMemory;
        return body;
    }

    // -----------------------------------------------------------------------
    // VTable implementations
    // -----------------------------------------------------------------------

    fn getTransactionImpl(ctx: *anyopaque, allocator: std.mem.Allocator, txid: []const u8) provider_mod.ProviderError!types.TransactionData {
        const self: *WhatsOnChainProvider = @ptrCast(@alignCast(ctx));

        const path = std.fmt.allocPrint(self.allocator, "/tx/hash/{s}", .{txid}) catch return provider_mod.ProviderError.OutOfMemory;
        defer self.allocator.free(path);

        const body = try self.httpGet(allocator, path);
        defer allocator.free(body);

        var parsed = std.json.parseFromSlice(std.json.Value, allocator, body, .{}) catch return provider_mod.ProviderError.NetworkError;
        defer parsed.deinit();

        const root = parsed.value.object;
        const tx_txid = if (root.get("txid")) |v| (if (v == .string) v.string else txid) else txid;
        const version: i32 = if (root.get("version")) |v| (if (v == .integer) @as(i32, @intCast(v.integer)) else 1) else 1;
        const locktime: u32 = if (root.get("locktime")) |v| (if (v == .integer) @as(u32, @intCast(v.integer)) else 0) else 0;

        // Parse inputs
        var inputs: []types.TxInput = &.{};
        if (root.get("vin")) |vin_val| {
            if (vin_val == .array) {
                const items = vin_val.array.items;
                var inp_list = allocator.alloc(types.TxInput, items.len) catch return provider_mod.ProviderError.OutOfMemory;
                for (items, 0..) |item, i| {
                    if (item == .object) {
                        const obj = item.object;
                        const in_txid = if (obj.get("txid")) |v| (if (v == .string) v.string else "") else "";
                        const vout: i32 = if (obj.get("vout")) |v| (if (v == .integer) @as(i32, @intCast(v.integer)) else 0) else 0;
                        var script_hex: []const u8 = "";
                        if (obj.get("scriptSig")) |ss| {
                            if (ss == .object) {
                                if (ss.object.get("hex")) |h| {
                                    if (h == .string) script_hex = h.string;
                                }
                            }
                        }
                        const seq: u32 = if (obj.get("sequence")) |v| (if (v == .integer) @as(u32, @intCast(v.integer)) else 0xffffffff) else 0xffffffff;
                        inp_list[i] = .{
                            .txid = allocator.dupe(u8, in_txid) catch return provider_mod.ProviderError.OutOfMemory,
                            .output_index = vout,
                            .script = allocator.dupe(u8, script_hex) catch return provider_mod.ProviderError.OutOfMemory,
                            .sequence = seq,
                        };
                    } else {
                        inp_list[i] = .{};
                    }
                }
                inputs = inp_list;
            }
        }

        // Parse outputs
        var outputs: []types.TxOutput = &.{};
        if (root.get("vout")) |vout_val| {
            if (vout_val == .array) {
                const items = vout_val.array.items;
                var out_list = allocator.alloc(types.TxOutput, items.len) catch return provider_mod.ProviderError.OutOfMemory;
                for (items, 0..) |item, i| {
                    if (item == .object) {
                        const obj = item.object;
                        // value is in BTC (float), convert to satoshis
                        var sats: i64 = 0;
                        if (obj.get("value")) |v| {
                            if (v == .float) {
                                sats = @intFromFloat(@round(v.float * 1e8));
                            } else if (v == .integer) {
                                sats = @intCast(v.integer);
                            }
                        }
                        var script_hex: []const u8 = "";
                        if (obj.get("scriptPubKey")) |sp| {
                            if (sp == .object) {
                                if (sp.object.get("hex")) |h| {
                                    if (h == .string) script_hex = h.string;
                                }
                            }
                        }
                        out_list[i] = .{
                            .satoshis = sats,
                            .script = allocator.dupe(u8, script_hex) catch return provider_mod.ProviderError.OutOfMemory,
                        };
                    } else {
                        out_list[i] = .{};
                    }
                }
                outputs = out_list;
            }
        }

        // Raw hex
        var raw: []const u8 = &.{};
        if (root.get("hex")) |v| {
            if (v == .string) raw = allocator.dupe(u8, v.string) catch return provider_mod.ProviderError.OutOfMemory;
        }

        return .{
            .txid = allocator.dupe(u8, tx_txid) catch return provider_mod.ProviderError.OutOfMemory,
            .version = version,
            .inputs = inputs,
            .outputs = outputs,
            .locktime = locktime,
            .raw = raw,
        };
    }

    fn broadcastImpl(ctx: *anyopaque, allocator: std.mem.Allocator, tx_hex: []const u8) provider_mod.ProviderError![]u8 {
        const self: *WhatsOnChainProvider = @ptrCast(@alignCast(ctx));

        // Build JSON body: {"txhex":"<raw tx hex>"}
        const json_body = std.fmt.allocPrint(self.allocator, "{{\"txhex\":\"{s}\"}}", .{tx_hex}) catch return provider_mod.ProviderError.OutOfMemory;
        defer self.allocator.free(json_body);

        const body = try self.httpPost(allocator, "/tx/raw", json_body);
        defer allocator.free(body);

        // WoC returns the txid as a JSON-encoded string (with quotes)
        // Strip surrounding quotes if present
        var result = body;
        if (result.len >= 2 and result[0] == '"' and result[result.len - 1] == '"') {
            result = result[1 .. result.len - 1];
        }

        return allocator.dupe(u8, result) catch return provider_mod.ProviderError.OutOfMemory;
    }

    fn getUtxosImpl(ctx: *anyopaque, allocator: std.mem.Allocator, address: []const u8) provider_mod.ProviderError![]types.UTXO {
        const self: *WhatsOnChainProvider = @ptrCast(@alignCast(ctx));

        const path = std.fmt.allocPrint(self.allocator, "/address/{s}/unspent", .{address}) catch return provider_mod.ProviderError.OutOfMemory;
        defer self.allocator.free(path);

        const body = try self.httpGet(allocator, path);
        defer allocator.free(body);

        var parsed = std.json.parseFromSlice(std.json.Value, allocator, body, .{}) catch return provider_mod.ProviderError.NetworkError;
        defer parsed.deinit();

        if (parsed.value != .array) {
            return allocator.alloc(types.UTXO, 0) catch return provider_mod.ProviderError.OutOfMemory;
        }

        const items = parsed.value.array.items;
        var result = allocator.alloc(types.UTXO, items.len) catch return provider_mod.ProviderError.OutOfMemory;
        for (items, 0..) |item, i| {
            if (item == .object) {
                const obj = item.object;
                const tx_hash = if (obj.get("tx_hash")) |v| (if (v == .string) v.string else "") else "";
                const tx_pos: i32 = if (obj.get("tx_pos")) |v| (if (v == .integer) @as(i32, @intCast(v.integer)) else 0) else 0;
                const value: i64 = if (obj.get("value")) |v| (if (v == .integer) @as(i64, @intCast(v.integer)) else 0) else 0;

                result[i] = .{
                    .txid = allocator.dupe(u8, tx_hash) catch return provider_mod.ProviderError.OutOfMemory,
                    .output_index = tx_pos,
                    .satoshis = value,
                    .script = allocator.dupe(u8, "") catch return provider_mod.ProviderError.OutOfMemory,
                };
            } else {
                result[i] = .{
                    .txid = allocator.dupe(u8, "") catch return provider_mod.ProviderError.OutOfMemory,
                    .output_index = 0,
                    .satoshis = 0,
                    .script = allocator.dupe(u8, "") catch return provider_mod.ProviderError.OutOfMemory,
                };
            }
        }

        return result;
    }

    fn getContractUtxoImpl(ctx: *anyopaque, allocator: std.mem.Allocator, script_hash: []const u8) provider_mod.ProviderError!?types.UTXO {
        const self: *WhatsOnChainProvider = @ptrCast(@alignCast(ctx));

        const path = std.fmt.allocPrint(self.allocator, "/script/{s}/unspent", .{script_hash}) catch return provider_mod.ProviderError.OutOfMemory;
        defer self.allocator.free(path);

        const body = self.httpGet(allocator, path) catch |err| {
            if (err == provider_mod.ProviderError.NotFound) return null;
            return err;
        };
        defer allocator.free(body);

        var parsed = std.json.parseFromSlice(std.json.Value, allocator, body, .{}) catch return provider_mod.ProviderError.NetworkError;
        defer parsed.deinit();

        if (parsed.value != .array) return null;
        const items = parsed.value.array.items;
        if (items.len == 0) return null;

        const first = items[0];
        if (first != .object) return null;
        const obj = first.object;

        const tx_hash = if (obj.get("tx_hash")) |v| (if (v == .string) v.string else "") else "";
        const tx_pos: i32 = if (obj.get("tx_pos")) |v| (if (v == .integer) @as(i32, @intCast(v.integer)) else 0) else 0;
        const value: i64 = if (obj.get("value")) |v| (if (v == .integer) @as(i64, @intCast(v.integer)) else 0) else 0;

        return .{
            .txid = allocator.dupe(u8, tx_hash) catch return provider_mod.ProviderError.OutOfMemory,
            .output_index = tx_pos,
            .satoshis = value,
            .script = allocator.dupe(u8, "") catch return provider_mod.ProviderError.OutOfMemory,
        };
    }

    fn getNetworkImpl(ctx: *anyopaque) []const u8 {
        const self: *WhatsOnChainProvider = @ptrCast(@alignCast(ctx));
        return self.network.toString();
    }

    fn getFeeRateImpl(_: *anyopaque) provider_mod.ProviderError!i64 {
        // BSV standard relay fee: 0.1 sat/byte = 100 sat/KB
        return 100;
    }

    fn getRawTransactionImpl(ctx: *anyopaque, allocator: std.mem.Allocator, txid: []const u8) provider_mod.ProviderError![]u8 {
        const self: *WhatsOnChainProvider = @ptrCast(@alignCast(ctx));

        const path = std.fmt.allocPrint(self.allocator, "/tx/{s}/hex", .{txid}) catch return provider_mod.ProviderError.OutOfMemory;
        defer self.allocator.free(path);

        const body = try self.httpGet(allocator, path);
        // Trim trailing whitespace/newlines
        var end: usize = body.len;
        while (end > 0 and (body[end - 1] == '\n' or body[end - 1] == '\r' or body[end - 1] == ' ')) {
            end -= 1;
        }
        if (end < body.len) {
            const trimmed = allocator.dupe(u8, body[0..end]) catch {
                allocator.free(body);
                return provider_mod.ProviderError.OutOfMemory;
            };
            allocator.free(body);
            return trimmed;
        }
        return body;
    }
};

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "WhatsOnChainProvider initializes correctly" {
    const allocator = std.testing.allocator;
    var woc = WhatsOnChainProvider.init(allocator, .mainnet);
    defer woc.deinit();

    const prov = woc.provider();
    try std.testing.expectEqualStrings("mainnet", prov.getNetwork());
}

test "WhatsOnChainProvider testnet URL" {
    const allocator = std.testing.allocator;
    var woc = WhatsOnChainProvider.init(allocator, .testnet);
    defer woc.deinit();

    const prov = woc.provider();
    try std.testing.expectEqualStrings("testnet", prov.getNetwork());

    const fee_rate = try prov.getFeeRate();
    try std.testing.expectEqual(@as(i64, 100), fee_rate);
}
